### PR TITLE
added dataset_size attribute to minari datasets

### PR DESCRIPTION
**Pull request by shreyansjainn**
# Description

added dataset size attribute to minari dataset generation function and cli. create a complimentary PR #143 as refactoring of code introduced lot of conflicts. 

Fixes # (issue), Depends on # (pull request)

## Type of change

> Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

### Screenshots

> Please attach before and after screenshots of the change if applicable.
> To upload images to a PR -- simply drag and drop or copy paste.

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have run `pytest -v` and no errors are present.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I solved any possible warnings that `pytest -v` has generated that are related to my code to the best of my knowledge.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by shreyansjainn** [comment]
hey @balisujohn @younik i've created this complimentary PR for #143 as there were a lot of conflicts due to code refactoring so i thought of doing the changes on the new structure from scratch. let me know if any more changes are required or not.

also, regarding the last change that John requested of getting the dataset size on the fly for a local dataset if the size is not present, i've not implemented that yet as i had to introduce a flag for `_show_dataset_table` method and it felt a bit cluttered. let me know if you'd like me to go ahead with that change or not. 

do let me know if any more changes are required in this.

**Review by younik — COMMENTED** [review]
Thanks for this PR; the code looks good, but I have a high-level doubt on how we want this to works.

The two possibilities are:
 - adding the size to the metadata, and just read it from it
 - Always "compute" it. It is not a real computation as the size is already stored in file metadata, but you have to iterate through files. 
 
For now you are doing the first option (however, in the get_dataset_size you are also computing for the cloud, that it would never use it as you call `get_dataset_size` only at the dataset creation).

I am personally more incline to the second simply because it is more robust to dataset changes however, it may slow down `minari list local` and `minari list remote`?
Can you maybe time the get_dataset_size method for datasets  with different number of file? (You can set a low max_episode_buffer on DataCollectorV0 to create datasets with many files)

File: minari/dataset/minari_storage.py [hunk 1]
```diff
@@ -403,3 +405,34 @@ def _add_episode_to_group(episode_buffer: Dict, episode_group: h5py.Group):
             episode_group.create_dataset(
                 key, data=data, dtype=dtype, chunks=True, maxshape=(None, *dshape)
             )
+
+
+def get_dataset_size(dataset_id: str):
```
[the remark this answers is not in the document]
Comment (younik) [reply minari/dataset/minari_storage.py:410]: (Depending on the high-level choice on the main comment, I will tell you were to place it)

[the remark this answers is not in the document]
Comment (shreyansjainn) [reply minari/dataset/minari_storage.py:410]: i can remove the gcp bucket part as anyway its not getting used, we are mainly calculating the size locally while creating the dataset only, so if i remove the gcp bucket part, even then should i move the method??

File: minari/dataset/minari_dataset.py [hunk 2]
```diff
@@ -138,6 +138,7 @@ def __init__(
         self._action_space = action_space
 
         self._generator = np.random.default_rng()
+        self._size = get_dataset_size(self._dataset_id)
```
[the remark this answers is not in the document]
Comment (shreyansjainn) [reply minari/dataset/minari_dataset.py:141]: sure, i'll remove it. i missed out on removing this.

**Comment by shreyansjainn** [comment]
@younik, have implemented some corrections. one thing left to correct is the way the dataset size is getting written in metadata for `create_dataset_from_collector_env`, so give me some time to correct that, will correct the unit tests accordingly too

**Comment by younik** [comment]
> @younik, have implemented some corrections. one thing left to correct is the way the dataset size is getting written in metadata for `create_dataset_from_collector_env`, so give me some time to correct that, will correct the unit tests accordingly too

Sounds good, let me know when you are already for another review

**Comment by shreyansjainn** [comment]
@younik have implemented the changes in both the create dataset methods and changed the unit tests accordingly so that it tests that action too. do let me know if any other change is needed.

File: minari/dataset/minari_storage.py [hunk 3]
```diff
@@ -403,3 +404,25 @@ def _add_episode_to_group(episode_buffer: Dict, episode_group: h5py.Group):
...
```
Comment (younik) [inline minari/dataset/minari_storage.py:409]: Okay, last thing before merging I think
Can you make this a class function of MInariStorage, rename it to `get_size` and remove the dataset_id arg?
So you can get the size using `storage.get_size()`

You can get the data_path from the MinariStorage attribute.

**Review by younik — CHANGES_REQUESTED** [review]

**Comment by younik** [comment]
Hey @shreyansjainn, what's the status of this? Do you need any help from my side?

**Comment by shreyansjainn** [comment]
@younik will work on this today or by tomorrow, was on break due to diwali festival in india and post that the work rush kept me occupied, will finish this asap...

File: tests/dataset/test_minari_storage.py [hunk 4]
```diff
@@ -170,3 +186,153 @@ def test_episode_metadata(tmp_dataset_dir):
...
+        author="WillDudley",
+        author_email="wdudley@farama.org",
+    )
+
+    # file_path = get_dataset_path(dataset_id)
+    # data_path = os.path.join(file_path, "data", "main_data.hdf5")
+    # original_dataset_size = os.path.getsize(data_path)
+    # original_dataset_size = np.round(original_dataset_size / 1000000, 1)
+
+    assert dataset._data.metadata['dataset_size'] == dataset.storage.get_size()
```
Comment (younik) [inline tests/dataset/test_minari_storage.py:330]: use `storage` instead of `_data`

File: tests/dataset/test_minari_storage.py [hunk 5]
```diff
@@ -170,3 +186,153 @@ def test_episode_metadata(tmp_dataset_dir):
...
+
+    # assert get_dataset_size(dataset_id) == original_dataset_size
+
+    check_data_integrity(dataset._data, dataset.episode_indices)
```
Comment (younik) [inline tests/dataset/test_minari_storage.py:334]: same

File: tests/dataset/test_minari_storage.py [hunk 6]
```diff
@@ -170,3 +186,153 @@ def test_episode_metadata(tmp_dataset_dir):
...
+        algorithm_name="random_policy",
+        code_permalink="https://github.com/Farama-Foundation/Minari/blob/f095bfe07f8dc6642082599e07779ec1dd9b2667/tutorials/LocalStorage/local_storage.py",
+        author="WillDudley",
+        author_email="wdudley@farama.org",
+    )
+
+    # file_path = get_dataset_path(dataset_id)
+    # data_path = os.path.join(file_path, "data", "main_data.hdf5")
+    # original_dataset_size = os.path.getsize(data_path)
+    # original_dataset_size = np.round(original_dataset_size / 1000000, 1)
```
Comment (younik) [inline tests/dataset/test_minari_storage.py:328]: remove unused code

File: tests/dataset/test_minari_storage.py [hunk 7]
```diff
@@ -170,3 +186,153 @@ def test_episode_metadata(tmp_dataset_dir):
...
+    dataset = minari.create_dataset_from_collector_env(
+        dataset_id=dataset_id,
+        collector_env=env,
+        algorithm_name="random_policy",
+        code_permalink="https://github.com/Farama-Foundation/Minari/blob/f095bfe07f8dc6642082599e07779ec1dd9b2667/tutorials/LocalStorage/local_storage.py",
+        author="WillDudley",
+        author_email="wdudley@farama.org",
+    )
+
+    assert dataset._data.metadata['dataset_size'] == dataset.storage.get_size()
```
Comment (younik) [inline tests/dataset/test_minari_storage.py:237]: use storage instead of _data

File: tests/dataset/test_minari_storage.py [hunk 8]
```diff
@@ -170,3 +186,153 @@ def test_episode_metadata(tmp_dataset_dir):
...
+
+    check_data_integrity(dataset._data, dataset.episode_indices)
```
Comment (younik) [inline tests/dataset/test_minari_storage.py:239]: same

**Review by younik — COMMENTED** [review]
Looks good to me, just if you can do a small cleanup on the test code as suggested

**Review by younik — APPROVED** [review]
LGTM, thanks!

**Pull request merged**